README.md

```diff
@@ -89,35 +109,49 @@ import { validate } from 'ember-validity-modifier';
 
 export default class MyForm extends Component {
   @action
-  async validateAndSubmit(event) {
-    let { target: form } = event;
-    event.preventDefault();
-    await validate(...form.elements);
-    if (form.reportValidity()) {
-      console.log('Fake submit action', Object.fromEntries(new FormData(form)));
-    }
+  validateAndSubmit({ target: form }) {
+    console.log('Fake submit action', Object.fromEntries(new FormData(form)));
   }
 }
 ```
 
-#### Example with `validator-update` event
+#### Example with `validateDidInsert` argument
 
-To validate the form state on initial render and any time its dependent arguments change, add the `'validator-update'` event to the list of events passed in via the `on` argument.
+To validate the form state on initial render add `validateDidInsert=true`.
 
 ```hbs
-  <input {{validity (fn this.matchTo this.match) on="change,validator-update"}}>
+  <input {{validity
+    (fn this.matchTo this.match)
+    on="change"
+    validateDidInsert=true
+  }}>
 ```
 
-```js
-export default MyComponent extends Component {
-  @action
-  matchTo(match, { value }) {
-    return value === match ? [] : ['Must match exactly'];
-  }
-}
+#### Example with `validateTracked` argument
+
+To validate the form state when  initial render and any time **one** of its dependent arguments change, add the `'watch'` argument with the dependent property.
+
+Because the validator argument is a function it is possible to not exercise the tracked properties and thus miss out on validations when those tracked properties change. This is the case of the `fn` helper which lazy executes thus doesn't trigger Ember's auto-tracking if it isn't ran first.
+
+To compensate we can use `validateTracked` to inform the modifier that it needs to run the validations when these properties change.
+
+```hbs
+  <input {{validity
+    (fn this.matchTo this.match)
+    on="change"
+    validateTracked=this.match
```

Comment:
This shouldn't be needed. As long as the modifier consumes all args when it's installed, they'll be entangled

Reply:
There is no way for the modifier to know what to consume as validations are lazy evaluated. So are helpers that return a function. The only way to have the modifier consume the properties is to list them explicitly which is what validatedTracked is all about.

Reply:
The only other option would be to force all validation functions to execute and ignore the results. But that would add a level of confusion to consumers to see their validator functions execute for no reason.

Reply:
oh interesting, I didn't know `fn` disconnects the reactivity :thinking: 

I'd probably switch to this api then:
https://github.com/NullVoxPopuli/tracked-modifier-demo/commit/1efc6f626f974f882a74169e7044e64767942863#diff-3537d24c7c7ac3a6699672257220a629b4f13066903e5b8b206097717f9d12e0R2

Reply:
`fn` seems to be lazy and will only register auto-tracking if it is executed.

Can you explain the demo api. I don't understand it. Also I expect this addon should remain backwards compatible to some older versions of Ember.

Reply:
that api is 3.25+, so it should be fine.

The gist is by _not_ using `fn`, the positional args' "trackedness" are consumed, so when `match` changes, the modifier tears itself down, and sets itself back up again -- eliminating the need in this PR to have `validateTracked`

Reply:
If I understand you are suggesting changing the `validity` modifier API to accept only *one* validator function and the rest of the params are curried args to the validator function. Thus triggering the auto-tracking?

I see where that goes but I know I and others have asked for the API that each params is it's own validator function. I presume others being familiar with the `(fn)` helper in that case. Or in that use of multiple helpers.

```hbs
<input {{validity
  (validate-something)
  (validate-something-else this.foobar)
}}>
```

Reply:
helpers should also auto-track like if those validators follow the function-factory pattern.

I imagine you could provide a wrapper function that each validator could use, like:

```js
export const validateSomething = validator((args) => {
  
});
```
where _validator_ does the entaglement from the demo

Reply:
@timlindvall What do you think about the idea that the act of adding the modifier will execute each validator function even if the result is ignored? If that seems like a reasonable trade off then I think I can accomplish @NullVoxPopuli proposal to drop the `validateTracked` API.

Reply:
>  will execute each validator function

I want to be _extra_ clear that this assumes that the validator functions would have to become factories that invoke the passed callback to `validator` when it's time to validate

Reply:
That's doable. I wonder if there is a way to tell the difference between a factory and a normal function. If so then we could have best of both worlds. 🤔

Reply:
you could use trickyJS and say `validator.__factory = true` and check for the presence of that property?

Reply:
Sorry, catching up on this thread. 😅 

It should be ok to have `{{validate}}` only accept one validation function per usage if we want to go that route. If someone wants multiple validators, they can attach multiple `{{validate}}` modifiers, which should collapse into one validation check. (And, if someone wants to pass in another validation method to a sub-component, composing using `...attributes` is a suitable option.)

Or, alternatively, the helper-function-factory pattern seemed to work well when I was working with the modifier, but I was using `validator-update` so it's possible that's because I was validating immediately. Though it would make sense to me that any props passed to a helper _should_ trigger autotracking. But... it carries the same caveat that you got at above... if I want to track something, I have to pass it in as an arg, even if I don't care about its value. (Perhaps that suggests a refactoring opportunity for that particular piece of code, though?)

Reply:
oh multiple `{{validate}}` should probably work :thinking: good thinking!

Reply:
I think there are two issues. Multiple modifiers on one element is already supported by this branch. Thus the restriction to only support one and let suffix params be args to the validator (partial application) is possible.

However, this won't help in the case of a user wishing to use the `(fn)` helper. the only way the `(fn)` helper will auto track is if the function is executed. Normally in the case of callbacks (actions up) this isn't really an issue. But it does become an issue if the function is not executed but deferred to another execution later like we do for validations since validators can be asynchronous.

We have three options I see

1. support `(fn)` via a `validateTracked` property —Needs documentation and might look odd
2. reinvent `(fn)` internally such that `[validator, ...argsForValidatorPartialApplication]` is the modifier's interface. —Users would be unaware of the caveats if they use `(fn)`
3. Provide mechanism for defining validator factories —adds extra complexity and still has the issue that users need to be aware of auto-tracking from a programming stand point.

I hate to say it but I'm still in the `validateTracked` camp. mainly because it is the least resistance to using the `(fn)` helper (or custom helpers without needing to exercise the auto-tracking that users will inevitably forget.

My understand so far is that the issue is that we are fighting two paradigms. Normally when something like a curried function is passed into something like a modifier it doesn't matter what it is. Only at execution time does it matter. In fact in the case of this modifier it also doesn't matter.

But it does matter if the tracked properties used to construct that function change and something unrelated (the act of validation) is dependent on that. Thus when you look at this situation both the validator *and* the modifier itself are dependents of the tracked properties.

By the design of tracked properties the thing that depends on them needs to use the. This is true for the `(fn)` helper and in our case true for the modifier itself.

That is the driving need for the `validateTracked` as it does the same thing as passing them as params though it is more explicit in this case. Here are two examples on how a modifier is intended to interact with tracked properties in terms of this addon:

#### Explicit

```hbs
<input {{validity (fn this.foo this.bar) validateTracked=(array this.foo this.bar)}}>
```

This version allows multiple validators in one modifier.

```hbs
<input {{validity
  (validate-foo this.foo)
  (validate-bar this.bar)
  validateTracked=(array this.foo this.bar)
}}>
```

#### Implicit

```hbs
<input {{validity (fn this.foo this.bar) this.foo this.bar}}>
```

This version forces only one validator per modifier.

```hbs
<input
  {{validity (validate-foo this.foo) this.foo}}
  {{validity (validate-bar this.bar) this.bar}}
>
```

Question is which one is easier better for users?

Reply:
What about a `tracked-validator` helper as a replacement for `fn` that actually exercises the passed args?

I'm experimenting with ideas. Or should we leave validateTracked as an optional API for those who like that syntax better?

This is a tough one as there really is only like three users of this addon. I don't know why good ideas don't spread.

Reply:
Imo, the best api is no api. Increased surface area is room for confusion.

Reply:
> Question is which one is easier better for users?

In your second example, you shouldn't need to duplicate the args

Reply:
```hbs
<input
  {{validity (validate-foo this.foo)}}
  {{validity (validate-bar this.bar)}}
>
```

Reply:
You are missing the point. You have to duplicate the args unless you can force all users who write custom helpers to exercise the args themselves. I think that ask is harder. And that users will easily be confused why the validation doesn't happen when their helper's args change. Also know that the act of validating is not by default when args change that would make things happen differently then normal forms do. It needs to be an opt-in situation.

Reply:
The only way to do this with deterministic results is to execute the validator functions in the modifier so that the auto-tracking gets exercised But that means validator functions can no longer be asynchronous which is a want by many who need validation. See ember-changeset-validations which use async validators.

Reply:
I also think it might be worth while to draft a RFC like proposal to expose some of the tracked primitives so that addons can tap into the internals maybe offer some kind of escape hatch for situations like this.

Reply:
> I think you are missing the point. You have to duplicate the args unless you can force all users who write custom helpers to exercise the args themselves. I

nay, I think I can maybe whip up a demo -- since worknig code is more effective than words.